### ISSUE TITLE
We're working on PGP support for this library

### ISSUE THREAD

**Issue by mmaridev** [ui/django-post_office#375]
Hi Post Office maintainers!

As per title of this issue, @dododevs and I are working on the implementation of the PGP standard for this library. The goal is being able to send PGP encrypted messages by passing the public key of the recipient to the `mail.send` method. We are focusing on [PGPy](https://github.com/SecurityInnovation/PGPy) since seems to be the most actively developed pure-Python PGP library.
Our goal is quite simple but we want to implement it in a way in which can be useful and, of course, upstream merged. Therefore, we came up with a few design questions/problems we wanted to discuss. The main theme is whether the public key of the recipient and a boolean flag to acknowledge if the mail needs to be encrypted or not are needed in the database. This kind of information could also be passed via the context field with the downside of resulting a dirtier solution.
Supporting PGP only when `render_on_delivery` is disabled would be much easier since all the computation could be done on the `mail.send` call, saving directly in the DB the encrypted message. This leads to another topic: should the (django) administrator be able to read encrypted mail or not? Cause if the answer is no, the solution described before is the only one possible. Otherwise, we should probably look into something stronger.

Watching forward for your direction,
Best,

**Comment by jrief** [comment]
In my opinion this feature makes a lot of sense. However, @selwin has to decide.

If implemented, I would create a model containing all recipients together with their public PGP key.
I also would encrypt the emails, just before they will be sent, not during creation. Reason is, that usually the administrator knows anyway what is going to be written. And this makes the use of templates much easier rather than storing the encrypted mail message in the database. Unless you have a use-case which really requires this.

**Comment by selwin** [comment]
I think this is a feature we can consider implementing. @jrief 's plan of attack also loooks good.

**Comment by jrief** [comment]
Something we maybe should consider, is to use a public PGP keyserver instead of a database model containing all public keys together with their email addresses.
Moreover, who would ever add that public key to the Post-Office database?
On the other side, public keyservers might be slow or unreliable, so caching the keys locally might make a lot of sense.

**Comment by mmaridev** [comment]
I agree with you @jrief: I am not totally against this feature but, in my opinion, should not be the default way to retrieve the key. This would introduce many problems: which PGP server we (or the sysadmin if configurable via settings) finds reliable and secure.
Given this considerations, on the application we are developing we'll be providing a user page linked to the relative model containing a field for the PGP public key. If the user wants he can retrieve it form a public keyserver using the email otherwise he can just simply paste it.
What do you think of adding such a feature just as a function in the utils? In this way we can say "ok, it's there, you can use it with caution".

**Comment by blag** [comment]
While I appreciate the interest and effort, you don't need to add PGP support directly to this project.

This can already be done using my django-secure-mail package ([PyPI](https://pypi.org/project/django-secure-mail/), [GitHub](https://github.com/blag/django-secure-mail)).

You should only need to set the [default backend of Post Office](https://github.com/ui/django-post_office?tab=readme-ov-file#custom-email-backends) to one of the offered [secure email backends](https://github.com/blag/django-secure-mail/blob/e2ffe48a40a69ba097a8841bd191bec103be00b6/secure_mail/backends.py#L166):

`settings.py`:
```python
POST_OFFICE = {
    ...
    'BACKENDS': {
        'default': 'secure_mail.backends.EncryptingSmtpEmailBackend',
        # or for development, one of these:
        # 'default': 'secure_mail.backends.EncryptingConsoleEmailBackend',
        # 'default': 'secure_mail.backends.EncryptingLocmemEmailBackend',
        # 'default': 'secure_mail.backends.EncryptingFilebasedEmailBackend',
    }
}
```

or if you want complete control over how your email backend works, you can define your own encrypting email backend using the [`EncryptingEmailBackendMixin`](https://github.com/blag/django-secure-mail/blob/e2ffe48a40a69ba097a8841bd191bec103be00b6/secure_mail/backends.py#L158) and set Post Office to use that instead:

`settings.py`:
```python
POST_OFFICE = {
    ...
    'BACKENDS': {
        'default': 'myapp.MyCustomEncryptingEmailBackend',
    }
}
```

I haven't touched the django-secure-mail project in a few years but it should still work in modern releases of Django (their email backend API hasn't changed in years). And if it doesn't, please open an issue or, even better, a PR for that project!

@ui  Since there is a valid, supported, and maintained workaround, I think this issue can be closed.
